pymeasure/instruments/keithley/keithleyDMM6500.py

```diff
@@ -0,0 +1,1432 @@
+        for the active mode, or can set another mode by its name.
+
+        :param mode: A valid :attr:`mode` name, or `None` for the active mode
+        """
+        self.write(f":SENS:{self._mode_command(mode)}:REL:STAT 1")
+
+    def disable_relative(self, mode=None):
+        """Disables the application of a relative offset value to the measurement
+        for the active mode, or can set another mode by its name.
+
+        :param mode: A valid :attr:`mode` name, or `None` for the active mode
+        """
+        self.write(f":SENS:{self._mode_command(mode)}:REL:STAT 0")
+
+    def acquire_relative(self, mode=None):
+        """Sets the active value as the relative for the active mode,
+        or can set another mode by its name.
+
+        :param mode: A valid :attr:`mode` name, or `None` for the active mode
+
+        :return: The relative value that was acquired
+
+        """
+        mode_cmd = self._mode_command(mode)
+        self.write(f":SENS:{mode_cmd}:REL:ACQ")
+        rel = float(self.ask(f":SENS:{mode_cmd}:REL?"))
+        return rel
+
+    def enable_filter(self, mode=None, type="repeat", count=1):
+        """Enables the averaging filter for the active mode,
+        or can set another mode by its name.
+
+        :param mode: A valid :attr:`mode` name, or `None` for the active mode
+        :param type: The type of averaging filter, could be ``REPeat``, ``MOVing``, or ``HYBRid``.
+        :param count: A number of averages, which can take take values from 1 to 100
+
+        :return: Filter status read from the instrument
+        """
+        mode_cmd = self._mode_command(mode)
+        self.write(f":SENS:{mode_cmd}:AVER:STAT 1")
+        self.write(f":SENS:{mode_cmd}:AVER:TCON {type}")
+        self.write(f":SENS:{mode_cmd}:AVER:COUN {count}")
+        return self.ask(f":SENS:{mode_cmd}:AVER:STAT?")
+
+    def disable_filter(self, mode=None):
+        """Disables the averaging filter for the active mode,
+        or can set another mode by its name.
+
+        :param mode: A valid :attr:`mode` name, or `None` for the active mode
+
+        :return: Filter status read from the instrument
+        """
+        mode_cmd = self._mode_command(mode)
+        self.write(f":SENS:{mode_cmd}:AVER:STAT 0")
+        return self.ask(f":SENS:{mode_cmd}:AVER:STAT?")
+
+    def beep(self, frequency, duration):
+        """Sounds a system beep.
+
+        :param frequency: A frequency in Hz between 20 Hz and 8000 Hz
+        :param duration: The amount of time to play the tone between 0.001 s to 100 s
+        :return: None
+        """
+        self.write(f":SYST:BEEP {frequency:g}, {duration:g}")
+
+    def write(self, command):
+        """Write a command to the instrument.
+
+        :param command: A command
+        :param type: str
+        :return: None
+        """
+        if "{function}" in command:
+            command = command.replace("{function}", KeithleyDMM6500.MODES[self.mode])
```

Comment:
Basically you do not need that if clause earlier on, as it only replaces, if it finds anything.

Alternatively, you can use also "format": `command.format(function=KeithleyDMM6500.MODES[self.mode])`, which will replace `{function}` with the appropriate value.


The same holds true for the Channel.

Reply:
Changed the code to use `string.format()`.

Reply:
You do not need that if clause, as format only replaces the variable, if encountered, see the examples:
~~~python
"x".format(a=5) == "x"
"x{a}".format(a=5) == "x5"
~~~

So you could write
~~~python
super().write(command.format(function=KeithleyDMM6500.MODES[self.mode]))
~~~
without the if clause, if you want to.

Reply:
It's better to keep if clause because when I removed it, I got `RecursionError`.

The reason is the `self.mode` will use `ask()` in the `common_base.py` to query instrument's mode but the step `self.write()` in the `ask()` function will call DMM6500's `write()` again and cause recursion.

Reply:
ah ok. Please add a comment to the code.

Reply:
Sure. Added a comment in the code.